lib/rust/flexer/src/lib.rs

```diff
@@ -14,26 +15,220 @@
+    definition: Definition
+}
+
+impl<Definition,Output,Reader> Deref for Flexer<Definition,Output,Reader>
+where Reader:LazyReader {
+    type Target = Definition;
+    fn deref(&self) -> &Self::Target {
+        &self.definition
+    }
+}
+
+impl<Definition,Output,Reader> DerefMut for Flexer<Definition,Output,Reader>
+where Reader:LazyReader {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.definition
+    }
+}
+
+impl<Definition,Output,Reader> Flexer<Definition,Output,Reader>
+where Definition: FlexerState<Reader>,Reader:LazyReader {
+    /// Creates a new lexer instance.
+    ///
+    /// Please note that the `reader` argument is currently hard-coded for testing purposes. This is
+    /// not the intention for the eventual design.
+    pub fn new(mut reader:Reader) -> Flexer<Definition,Output,Reader> {
+        let mut state_stack = Vec::new();
+        state_stack.reserve(1024);
+        let current_match = String::from("");
+        let status = FlexerStageStatus::Initial;
+        let mut tokens = Vec::new();
+        tokens.reserve(1024);
+        let definition = Definition::new(&mut reader);
+        let initial_state_id = definition.initial_state().id;
+        state_stack.push(initial_state_id);
+
+        Flexer {state_stack,reader,current_match,status,tokens,definition}
+    }
+}
+
+/// This block is things that are part of the lexer's interface and functionality.
+impl<Definition,Reader,Output> Flexer<Definition,Output,Reader>
+where Definition:FlexerState<Reader>,Reader:LazyReader,Output:Clone {
+    /// Gets the lexer result.
+    pub fn get_result(&mut self) -> Option<Vec<Output>> {
+        Some(self.tokens.clone())
+    }
+
+    /// Gets the lexer's root state.
+    pub fn root_state(&self) -> &Group {
+        &self.definition.initial_state()
+    }
+
+    /// Gets the state that the lexer is currently in.
+    pub fn current_state(&self) -> usize {
+        *self.state_stack.last().expect("There should always be one state on the stack.")
+    }
+
+    /// Tells the lexer to enter the state described by `state`.
+    pub fn begin_state(&mut self,state:usize) {
+        self.state_stack.push(state);
+    }
+
+    /// Ends the current state, returning the popped state identifier if one was ended.
+    pub fn end_state(&mut self) -> Option<usize> {
+        if self.state_stack.len() > 1 {
+            let ix = self.state_stack.pop().expect("There should be an item to pop.");
+            Some(ix)
+        } else {
+            None
+        }
+    }
+
+    /// Ends states until the specified `state` is reached, leaving the lexer in `state`.
+    ///
+    /// If `state` does not exist on the lexer's stack, then the lexer will be left in the root
+    /// state.
+    pub fn end_states_until(&mut self,state:usize) -> Vec<usize> {
+        // Never drop the root state
+        let position_of_target =
+            self.state_stack.iter().positions(|elem| *elem == state).last().unwrap_or(0);
```

Comment:
instead of comment distribute it over few well named variables, like `let opt_non_root_state = ...` etc.